partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt

```diff
@@ -241,16 +242,88 @@ internal class PlanTyper(
+            if (!setOpSchemaTypesMatch(lhs, rhs)) {
+                return createRelErrForSetOpMismatchTypes()
+            }
+            // Compute Schema
+            val type = Rel.Type(lhs.type.schema, props = emptySet())
+            return Rel(type, node.copy(lhs = lhs, rhs = rhs))
         }
 
         override fun visitRelOpIntersect(node: Rel.Op.Intersect, ctx: Rel.Type?): Rel {
-            TODO("Type RelOp Intersect")
+            val lhs = visitRel(node.lhs, node.lhs.type)
+            val rhs = visitRel(node.rhs, node.rhs.type)
+            // Check for Compatibility
+            if (!setOpSchemaSizesMatch(lhs, rhs)) {
+                return createRelErrForSetOpMismatchSizes()
+            }
+            if (!setOpSchemaTypesMatch(lhs, rhs)) {
+                return createRelErrForSetOpMismatchTypes()
+            }
+            // Compute Schema
+            val type = Rel.Type(lhs.type.schema, props = emptySet())
+            return Rel(type, node.copy(lhs = lhs, rhs = rhs))
         }
 
-        override fun visitRelOpExcept(node: Rel.Op.Except, ctx: Rel.Type?): Rel {
-            TODO("Type RelOp Except")
+        override fun visitRelOpUnion(node: Rel.Op.Union, ctx: Rel.Type?): Rel {
+            val lhs = visitRel(node.lhs, node.lhs.type)
+            val rhs = visitRel(node.rhs, node.rhs.type)
+            // Check for Compatibility
+            if (!setOpSchemaSizesMatch(lhs, rhs)) {
+                return createRelErrForSetOpMismatchSizes()
+            }
+            if (!setOpSchemaTypesMatch(lhs, rhs)) {
+                return createRelErrForSetOpMismatchTypes()
+            }
+            // Compute Schema
+            val size = max(lhs.type.schema.size, rhs.type.schema.size)
+            val schema = List(size) {
+                val lhsBinding = lhs.type.schema.getOrNull(it) ?: Rel.Binding("_$it", MISSING)
+                val rhsBinding = rhs.type.schema.getOrNull(it) ?: Rel.Binding("_$it", MISSING)
+                val bindingName = when (lhsBinding.name == rhsBinding.name) {
+                    true -> lhsBinding.name
+                    false -> "_$it"
+                }
+                Rel.Binding(bindingName, unionOf(lhsBinding.type, rhsBinding.type).flatten())
+            }
+            val type = Rel.Type(schema, props = emptySet())
+            return Rel(type, node.copy(lhs = lhs, rhs = rhs))
+        }
+
+        /**
+         * @return whether each type of the [lhs] is comparable to its counterpart on the [rhs]
+         * @param lhs should be typed already
+         * @param rhs should be typed already
+         */
+        private fun setOpSchemaTypesMatch(lhs: Rel, rhs: Rel): Boolean {
+            // TODO: [RFC-0007](https://github.com/partiql/partiql-lang/blob/main/RFCs/0007-rfc-bag-operators.md)
+            //  states that the types must be "comparable". For now, we will always return true. In the future, we need
+            //  to add support for checking comparable types.
```

Comment:
For now, do not do comparability check of types. Only check same number of columns for both set ops.